### ISSUE TITLE
[Discussion] Tox Modules/Extensions Framework

### ISSUE THREAD

**Issue by thyseus** [irungentoo/toxcore#1259]
# Prolog
I hope this is the correct place for this sort of discussion; if not - please tell me, where it is instead.

# Introduction
Being inspired from the idea of an truly peer to peer text/audio/voice messenger like tox, i came to the idea that "the next big thing" could be a decentralized, end-to-end encrypted social network running on semantic technology (http://www.w3.org/RDF/, http://www.obitko.com/tutorials/ontologies-semantic-web/rdf-graph-and-syntax.html), giving the user total freedom about what to share with whom. Much like the promises Diaspora had (https://diasporafoundation.org/), but totally decentralized without the need of semi-central Pods that aggregate your information; but rather use a technology like bittorrent sync (http://www.getsync.com/intl/de/) to distribute your information.

# tl;dr version
Before i elaborate my thoughts on this even more, here is the tl;dr Version of my [Discussion]-Issue:

Are there any plans for providing a module or extension framework for the tox-core protocol? So developers can massage in their additional features without the need to fork the whole project? I can imagine a third-party client would easily exhaust its possibilites when creating more sophisticated
applications on the tox-core library.

# blueprint of a truly decentralized social network on top of the tox-core library
A user installs the Software - lets call it "Spider", having all the features of a social network except the messaging features (which tox itself already is covering very well), like a user Timeline, Feed, Friendship, Events, Inviations, etc. in an offline Application. This Application looks like qTox/µTox
Everything done in this application is done Offline, saving all the Activities inside a bunch of .rdf files. Using the Turtle Format as an example ( https://en.wikipedia.org/wiki/Turtle_(syntax) ), this file(s) could look like:

```
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix pf: <http://tox.im/spider/1.0/profile> .
@prefix ev: <http://tox.im/spider/1.0/events> .
@prefix fe: <http://tox.im/spider/1.0/feed> .
@prefix fs: <http://tox.im/spider/1.0/friendship> .

:me a pf:Person ;
	    me:unique_id "4468e5deabf5e6d0740cd1a77df56f67093ec943"
            me:givenname "Jacque" ;
            me:family_name "Fresco" ;
            me:homepage <https://www.thevenusproject.com/en/> ;
            me:mbox <mailto:foo@bar.org> .

:ev a ev:Event;
	ev:name "My Party" ;
	ev:date "2015-02-16" ;

:ev a ev:Event;
	ev:name "My Birthday" ;
	ev:date "1984-02-16" ;
	ev:recurral "yearly" ;

fs: a fs:Friendship;
	ev:with "c21bbeffa830a0290498a41b6be8a3c29974df64"
	ev:since "2000-01-01"

fs: a fs:Friendship;
	ev:with "c8aa1aeedcc505699bc165d7c7b193fa0f7e62aa"
	ev:since "2005-01-01"

```

All his Information is available on his Computer, for offline availability, backupability, and so on. Now, every piece of Information that i gather here is being synced encrypted using DHT and bittorrent sync (or a to-be-created protocol optimized for this sort of storage). Inside the GUI i can choose which information is public - it would be encrypted with only my private key, so everyone can see this information (ensuring author,- and ownership of the information). Everything else is being encrypted with my private key, that only the people in my friend list (or whoever i provide it) 
can see.

Changes made to my profile information or posting new Things on my Timeline would be propagated automatically using the DHT network. All my Friends would get the new information first, due to the close distance. They would propagate the changes to their peers again, until everything is in sync. This technology would remove the central master Hub where all the data is collected, but makes the information redundant and available for everyone with access privileges.

# Conclusion

Maybe there already is a project implementing stuff like this. I haven´t found one yet going in this direction; Diaspora is still "to central" with its pod technology for my taste. Please let me know if you know any sort of project doing this. Would the tox-core library be a good technology for
implementing this sort of very-decentralized social network? What´s your ideas?

Oh, and - thanks for this great - and (sadly) in our political reality extremely important(!!) - Software. Keep up the good work...

**Comment by thyseus** [comment]
I have found https://github.com/AKSW/xodx and http://www.semantic-web-journal.net/sites/default/files/swj201_4.pdf which look a lot like what i am proposing. I am not sure if their focus is on strong cryptography like tox has. Is a Implementation of this on the basis of the toxcore library feasible? Whats your opinion?

**Comment by czarkoff** [comment]
@thyseus

> Is a Implementation of this on the basis of the toxcore library feasible?

The best way to answer this question is to try implementing the project. Honor is yours.

FWIW unless I had diffs, I would post my proposal on wiki, not on bug tracker.

**Comment by drequivalent** [comment]
I think, something much like Jabber's XEP body of standards is in order.

**Comment by thyseus** [comment]
I have found https://www.ethereum.org/ which is looks very promising for the project i want to undertake.

**Comment by GrayHatter** [comment]
I didn't read too deeply, but unless I missed something, toxcore already supports this. All you need to do is write a client.

**Comment by ProMcTagonist** [comment]
Please link your project here when you start.
